File: nxflix-agents-python/src/nxflix_agents/routers/study.py
"""Study session API routes."""

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel

from nxflix_agents.models import (
    RecommendationRequest,
    SessionRequest,
    SessionCompleteRequest,
    StudyRecommendation,
    StudySession,
    SessionResult,
    UserProgress,
)
from nxflix_agents.agents import StudyOrchestratorAgent
from nxflix_agents.services import GrammarService, SM2Service

router = APIRouter(prefix="/api/study", tags=["study"])

# In-memory storage for demo (replace with database)
_user_progress: dict[str, dict[str, UserProgress]] = {}

# Singleton services (would be injected in production)
_grammar_service = GrammarService()
_sm2_service = SM2Service()
_study_agent = StudyOrchestratorAgent(_grammar_service, _sm2_service)


def get_study_agent() -> StudyOrchestratorAgent:
    """Dependency to get the study orchestrator agent."""
    return _study_agent


def get_user_progress(user_id: str) -> list[UserProgress]:
    """Get user progress for all grammar points."""
    return list(_user_progress.get(user_id, {}).values())


class RecommendationsResponse(BaseModel):
    """Response for study recommendations."""

    recommendation: StudyRecommendation


class SessionResponse(BaseModel):
    """Response for session operations."""

    session: StudySession


class SessionCompleteResponse(BaseModel):
    """Response for completing a session."""

    session: StudySession
    updated_grammar_ids: list[str]


@router.post("/recommendations", response_model=RecommendationsResponse)
async def get_recommendations(
    request: RecommendationRequest,
    agent: StudyOrchestratorAgent = Depends(get_study_agent),
):
    """Get personalized study recommendations."""
    user_progress = get_user_progress(request.user_id)

    recommendation = await agent.get_recommendations(request, user_progress)

    return RecommendationsResponse(recommendation=recommendation)


@router.post("/sessions", response_model=SessionResponse)
async def start_session(
    request: SessionRequest,
    agent: StudyOrchestratorAgent = Depends(get_study_agent),
):
    """Start a new study session."""
    session = await agent.start_session(request)
    return SessionResponse(session=session)


@router.put("/sessions/{session_id}/complete", response_model=SessionCompleteResponse)
async def complete_session(
    session_id: str,
    request: SessionCompleteRequest,
    agent: StudyOrchestratorAgent = Depends(get_study_agent),
):
    """Complete a study session and update progress."""
    # Parse results
    results = [
        SessionResult(
            grammar_id=r.get("grammar_id", ""),
            questions_asked=r.get("questions_asked", 0),
            correct_answers=r.get("correct_answers", 0),
            score=r.get("score", 0.0),
        )
        for r in request.results
    ]

    # Get current user progress
    session = agent.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    user_id = session.user_id
    user_progress_dict = _user_progress.get(user_id, {})

    # Create progress entries for new grammar points
    for result in results:
        if result.grammar_id not in user_progress_dict:
            user_progress_dict[result.grammar_id] = UserProgress(
                user_id=user_id,
                grammar_id=result.grammar_id,
            )

    # Complete the session
    completed_session, updated_progress = await agent.complete_session(
        session_id, results, user_progress_dict
    )

    # Update stored progress
    if user_id not in _user_progress:
        _user_progress[user_id] = {}
    _user_progress[user_id].update(updated_progress)

    return SessionCompleteResponse(
        session=completed_session,
        updated_grammar_ids=list(updated_progress.keys()),
    )


@router.get("/sessions/{session_id}", response_model=SessionResponse)
async def get_session(
    session_id: str,
    agent: StudyOrchestratorAgent = Depends(get_study_agent),
):
    """Get a study session by ID."""
    session = agent.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return SessionResponse(session=session)

File: nxflix-agents-ts/src/routers/study.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { StudyOrchestratorAgent } from '../agents/study-orchestrator.js';
import type { SessionResult, UserProgress } from '../models/progress.js';
import { userProgress, grammarService, sm2Service, getUserProgressList } from '../state.js';

const studyRouter = Router();

// Use shared services from state
const studyAgent = new StudyOrchestratorAgent(grammarService, sm2Service);

// Request schemas
const RecommendationRequestSchema = z.object({
  userId: z.string(),
  availableGrammarIds: z.array(z.string()).default([]),
  maxItems: z.number().int().min(1).max(20).default(5),
  focusWeakAreas: z.boolean().default(true),
  includeNew: z.boolean().default(true),
  timeAvailableMinutes: z.number().int().nullish(),
});

const SessionRequestSchema = z.object({
  userId: z.string(),
  grammarIds: z.array(z.string()),
  questionCount: z.number().int().min(1).max(50).default(10),
  questionTypes: z.array(z.string()).default([]),
  difficulty: z.number().int().min(1).max(5).nullish(),
});

const SessionCompleteRequestSchema = z.object({
  sessionId: z.string(),
  results: z.array(z.object({
    itemId: z.string(),
    contentType: z.enum(['grammar', 'vocabulary', 'reading', 'kanji', 'listening']).default('grammar'),
    questionsAsked: z.number().int(),
    correctAnswers: z.number().int(),
    score: z.number(),
  })),
});

// POST /api/study/recommendations
studyRouter.post('/recommendations', async (req: Request, res: Response) => {
  try {
    const request = RecommendationRequestSchema.parse(req.body);
    const progress = getUserProgressList(request.userId);

    const recommendation = await studyAgent.getRecommendations(request, progress);

    res.json({ recommendation });
  } catch (error) {
    console.error('Error getting recommendations:', error);
    res.status(400).json({ error: String(error) });
  }
});

// POST /api/study/sessions
studyRouter.post('/sessions', async (req: Request, res: Response) => {
  try {
    const request = SessionRequestSchema.parse(req.body);
    const session = await studyAgent.startSession(request);

    res.json({ session });
  } catch (error) {
    console.error('Error starting session:', error);
    res.status(400).json({ error: String(error) });
  }
});

// PUT /api/study/sessions/:id/complete
studyRouter.put('/sessions/:id/complete', async (req: Request<{ id: string }>, res: Response) => {
  try {
    const sessionId = req.params.id;
    const { results } = SessionCompleteRequestSchema.parse({
      ...req.body,
      sessionId,
    });

    const session = studyAgent.getSession(sessionId);
    if (!session) {
      res.status(404).json({ error: 'Session not found' });
      return;
    }

    const userId = session.userId;
    const userProgressDict: Record<string, UserProgress> = userProgress[userId] ?? {};

    // Create progress entries for new items
    for (const result of results) {
      const progressKey = `${result.contentType}:${result.itemId}`;
      if (!(progressKey in userProgressDict)) {
        userProgressDict[progressKey] = {
          userId,
          itemId: result.itemId,
          contentType: result.contentType,
          sm2Data: { easeFactor: 2.5, interval: 0, repetitions: 0 },
          timesStudied: 0,
          timesCorrect: 0,
          lastScore: null,
          masteryLevel: 0,
        };
      }
    }

    const sessionResults: SessionResult[] = results.map(r => ({
      itemId: r.itemId,
      contentType: r.contentType,
      questionsAsked: r.questionsAsked,
      correctAnswers: r.correctAnswers,
      score: r.score,
    }));

    const { session: completedSession, updatedProgress } = await studyAgent.completeSession(
      sessionId,
      sessionResults,
      userProgressDict
    );

    // Update stored progress
    if (!userProgress[userId]) {
      userProgress[userId] = {};
    }
    Object.assign(userProgress[userId], updatedProgress);

    res.json({
      session: completedSession,
      updatedGrammarIds: Object.keys(updatedProgress),
    });
  } catch (error) {
    console.error('Error completing session:', error);
    res.status(400).json({ error: String(error) });
  }
});

// GET /api/study/sessions/:id
studyRouter.get('/sessions/:id', (req: Request<{ id: string }>, res: Response) => {
  const session = studyAgent.getSession(req.params.id);
  if (!session) {
    res.status(404).json({ error: 'Session not found' });
    return;
  }
  res.json({ session });
});

export { studyRouter };
